### PR TITLE
Object Storage Sink

### PR DESCRIPTION
**Pull request by lavkesh**

### CODE REVIEW

File: src/main/java/io/odpf/firehose/config/converter/CloudSinkPartitioningTypeConverter.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package io.odpf.firehose.config.converter;
+
+import io.odpf.firehose.sink.objectstorage.Constants;
+import org.aeonbits.owner.Converter;
+
+import java.lang.reflect.Method;
+
+public class CloudSinkPartitioningTypeConverter implements Converter<Constants.PartitioningType> {
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/config/converter/CloudSinkPartitioningTypeConverter.java:8]: Add file level documentation as part of every file

File: src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    long getFileRotationDurationMillis();
+
+    @Key("SINK_OBJECT_STORAGE_ROTATION_MAX_SIZE_BYTES")
+    @DefaultValue("268435456")
+    long getFileRationMaxSizeBytes();
+
+    @Key("SINK_OBJECT_STORAGE_TIME_PARTITIONING_FIELD_NAME")
+    String getTimePartitioningFieldName();
+
+    @Key("SINK_OBJECT_STORAGE_TIME_PARTITIONING_DATE_PATTERN")
```
[the remark this answers is not in the document]
Comment (fzrvic) [reply src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java:43]: an example date partitioning without dash, like `/dt=20210101`

File: src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Key("SINK_OBJECT_STORAGE_TIME_PARTITIONING_HOUR_PREFIX")
+    @DefaultValue("hr=")
+    String getTimePartitioningHourPrefix();
+
+    @Key("SINK_OBJECT_STORAGE_TYPE")
+    @DefaultValue("GCS")
+    String getCloudStorageType();
+
+    @Key("SINK_OBJECT_STORAGE_BUCKET_NAME")
```
[the remark this answers is not in the document]
Comment (fzrvic) [reply src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java:68]: currently we are using any secret that path defined on `GOOGLE_APPLICATION_CREDENTIALS` variable

File: src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java:43]: removed this.

File: src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java:68]: we will add a config for auth, which can be supplied via kube secrets.

File: src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    @DefaultValue("1048576")
+    int getWriterPageSize();
+
+    @Key("SINK_OBJECT_STORAGE_ROTATION_DURATION_MILLIS")
+    @DefaultValue("3600000")
+    long getFileRotationDurationMillis();
+
+    @Key("SINK_OBJECT_STORAGE_ROTATION_MAX_SIZE_BYTES")
+    @DefaultValue("268435456")
+    long getFileRationMaxSizeBytes();
```
[the remark this answers is not in the document]
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/config/ObjectStorageSinkConfig.java:38]: fixed it

File: src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSinkFactory.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class ObjectStorageSinkFactory implements SinkFactory {
+    @Override
+    public Sink create(Map<String, String> configuration, StatsDReporter statsDReporter, StencilClient stencilClient) {
+        ObjectStorageSinkConfig sinkConfig = ConfigFactory.create(ObjectStorageSinkConfig.class, configuration);
```
[the remark this answers is not in the document]
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSinkFactory.java:35]: refactored.

File: src/main/java/io/odpf/firehose/sink/objectstorage/message/MessageSerializer.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.gojek.de.stencil.parser.Parser;
+import com.google.protobuf.DynamicMessage;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.odpf.firehose.consumer.Message;
+import io.odpf.firehose.exception.DeserializerException;
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class MessageSerializer {
```
[the remark this answers is not in the document]
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/sink/objectstorage/message/MessageSerializer.java]: fixed it

File: src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSink.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            this.writerOrchestrator.write(record);
+        }
+        return new LinkedList<>();
+    }
+
+    @Override
+    protected void prepare(List<Message> messages) throws DeserializerException, IOException, SQLException {
+        records = new LinkedList<>();
+        for (Message message : messages) {
+            Record record = this.writerOrchestrator.convertToRecord(message);
```
[the remark this answers is not in the document]
Comment (lavkesh) [reply src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSink.java]: fixed it.

**Pull request merged**